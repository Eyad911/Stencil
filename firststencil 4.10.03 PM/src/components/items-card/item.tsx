export interface Item {
    id: string;
  label: string;
  price: {
    amount: number;
    currency: string;
  };
  qty: number;
  thumbnail: string;
    }
