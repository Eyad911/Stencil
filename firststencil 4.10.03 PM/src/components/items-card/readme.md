# ey-items-card Component

The ey-items-card component is a Stencil web component designed to display a list of items in a card format. Each item includes a thumbnail image, label, price, and quantity. This component is flexible, allowing the input data to be provided either as an array of objects or a JSON string.

## Table of Contents

- Properties
- Watchers
- Lifecycle Methods
- Rendering
- Example Usage
- Item Interface
- Styling
- Additional Information

## Properties

### arrayData

- Type: Item[] | string
- **Description** : The data source for the component. It can be either an array of Item objects or a JSON string representing the array of Item objects.

## Watchers

### arrayDataWatcher

- **Description** : This watcher method is triggered whenever the arrayData property changes. It checks the type of the new value and, if it is a JSON string, parses it into an array of Item objects and assigns it to the private _arrayData property. If it is already an array of Item objects, it directly assigns it to _arrayData.

## Lifecycle Methods

### componentWillLoad

- **Description** : This lifecycle method is called just before the component is rendered. It ensures the arrayData is correctly parsed and assigned by calling the arrayDataWatcher method.

## Rendering

The render method generates the HTML structure for the component. Each item in the _arrayData array is displayed with its thumbnail, label, price, and quantity.

### HTML Structure

- **Container** : The outermost div has several CSS classes for styling, including padding, background color, and border.

- **Item** : Each item is wrapped in a div with padding. It includes:
 - **Thumbnail** : An (img) element displaying the item's thumbnail image.
 - **Details** :  A (div) containing the item's label and price.
 - **Quantity and Total Price** : A (div) displaying the item's quantity and total price (calculated as quantity multiplied by price).

## Styling

The component uses Tailwind CSS classes for styling, including spacing, colors, font styles, and more.

## Example Usage

```
 <ey-items-card
      array-data='[
{
  "id": "001",
  "label": "JBL QuietComfort 45 wireless bluetooth headphones",
  "thumbnail": "https://cdn.salla.sa/oRmZv/bi7pdht4SRGPLmziP5fDcuFyuDOwz4FypJ5svKxU.jpg",
  "qty": 1,
  "price": {        
    "currency": "SAR",
    "amount": 1500
  }      
},
{
  "id": "002",
  "label": "JBL Active Noise Cancelling Headphones",
  "thumbnail": "https://cdn.salla.sa/oRmZv/9wgo7gyYLL6WsYwugU1zZEn3nypaI93MDCPiPhI5.png",
  "qty": 1,
  "price": {        
    "currency": "SAR",
    "amount": 750
  }     
},
]'
    ></ey-items-card>

```

## Item Interface

```
export interface Item {
    id: string;
  label: string;
  price: {
    amount: number;
    currency: string;
  };
  qty: number;
  thumbnail: string;
    }
```

## Styling

To customize the styles, you can modify the items-card.css file or adjust the Tailwind CSS classes used in the render method.

## Additional Information

For more details on using Stencil components, refer to the [Stencil documentation](https://stenciljs.com/docs/introduction). This will provide you with comprehensive information on how to create and work with Stencil components effectively.





<!-- Auto Generated Below -->


## Properties

| Property    | Attribute    | Description | Type               | Default     |
| ----------- | ------------ | ----------- | ------------------ | ----------- |
| `arrayData` | `array-data` |             | `Item[] \| string` | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
