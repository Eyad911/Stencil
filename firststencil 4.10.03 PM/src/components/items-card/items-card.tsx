import { Component, Prop, h, Watch } from '@stencil/core';
import { Item } from './item';

@Component({
  tag: 'ey-items-card',
  styleUrl: 'items-card.css',
  shadow: true,
})
export class ItemsCard {
  private _arrayData: Item[] = [];

  @Prop() arrayData: Item[] | string;

  @Watch('arrayData')
  arrayDataWatcher(newValue: Item[] | string) {
    if (typeof newValue === 'string') {
      try {
        this._arrayData = JSON.parse(newValue) || [];
      } catch (error) {
        // console.error('Invalid JSON string provided for arrayData', error);
        this._arrayData = [];
      }
    } else if (Array.isArray(newValue)) {
      this._arrayData = newValue;
    } else {
      this._arrayData = [];
    }
  }

  componentWillLoad() {
    this.arrayDataWatcher(this.arrayData);
  }

  render() {
    return (
      <div class="w-full px-[8px] bg-[#baf3e6] bg-opacity-10 border-[2px] border-[#BAF3E626] rounded-[6px]">
        {this._arrayData.map((item, index) => (
          <div key={index} class="py-[10px]">
            <div class="flex space-x-8">
              <div class="flex space-x-3">
                <img src={item.thumbnail} class="w-[45px] h-[45px] rounded-[22.5px]" alt={item.label} />
                <div class="w-[225px]">
                  <p class="text-[#004d5a] font-normal text-xs/[16px] underline">{item.label}</p>
                  <p class="mt-[4px] text-[#999999] opacity-90 text-[10px]/[16px] space-x-1">
                    {item.price.currency} {item.price.amount}
                  </p>
                </div>
              </div>
              <div class="flex justify-between text-[#333333] text-xs/[16px] font-bold w-full items-center p-2">
                <p>{item.qty.toString()}</p>
                <p>
                  {item.price.currency} {item.price.amount.toFixed(2)}
                </p>
              </div>
            </div>
            {index !== this._arrayData.length - 1 && (
              <div
                style={{
                  borderTop: '2px solid rgba(186, 243, 230, 1)',
                  marginTop: '20px',
                }}
              />
            )}
          </div>
        ))}
      </div>
    );
  }
}
