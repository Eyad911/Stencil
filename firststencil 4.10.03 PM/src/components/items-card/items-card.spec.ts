import { newSpecPage } from '@stencil/core/testing';
import { ItemsCard } from './items-card';
import { Item } from './item';

describe('ItemsCard', () => {
  it('should build', () => {
    expect(new ItemsCard()).toBeTruthy();
  });

  it('should render correctly with arrayData', async () => {
    const page = await newSpecPage({
      components: [ItemsCard],
      html: `<ey-items-card></ey-items-card>`,
    });

    let component = page.rootInstance as ItemsCard;
    component.arrayData = [
      {
        thumbnail: 'https://example.com/image.jpg',
        label: 'Test Item',
        price: { currency: 'USD', amount: 100 },
        qty: 1,
      },
    ] as Item[];

    component.arrayDataWatcher(component.arrayData);

    await page.waitForChanges();

    expect(page.root).toEqualHtml(`
      <ey-items-card>
        <div class="w-full px-[8px] bg-[#baf3e6] bg-opacity-10 border-[2px] border-[#BAF3E626] rounded-[6px]">
          <div class="py-[10px]">
            <div class="flex space-x-8">
              <div class="flex space-x-3">
                <img src="https://example.com/image.jpg" class="w-[45px] h-[45px] rounded-[22.5px]" alt="Test Item" />
                <div class="w-[225px]">
                  <p class="text-[#004d5a] font-normal text-xs/[16px] underline">Test Item</p>
                  <p class="mt-[4px] text-[#999999] opacity-90 text-[10px]/[16px] space-x-1">USD 100</p>
                </div>
              </div>
              <div class="flex justify-between text-[#333333] text-xs/[16px] font-bold w-full items-center p-2">
                <p>1</p>
                <p>USD 100.00</p>
              </div>
            </div>
          </div>
        </div>
      </ey-items-card>
    `);
  });
});