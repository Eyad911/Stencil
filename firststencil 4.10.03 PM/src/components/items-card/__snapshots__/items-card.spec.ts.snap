// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ey-items-card should render correctly with arrayData as a string 1`] = `
<ey-items-card>
  <mock:shadow-root>
    <div class="bg-[#baf3e6] bg-opacity-10 border-[#BAF3E626] border-[2px] px-[8px] rounded-[6px] w-full">
      <div class="py-[10px]">
        <div class="flex space-x-8">
          <div class="flex space-x-3">
            <img alt="Item 1" class="h-[45px] rounded-[22.5px] w-[45px]" src="https://via.placeholder.com/45">
            <div class="w-[225px]">
              <p class="font-normal text-[#004d5a] text-xs/[16px] underline">
                Item 1
              </p>
              <p class="mt-[4px] opacity-90 space-x-1 text-[#999999] text-[10px]/[16px]">
                USD 12.5
              </p>
            </div>
          </div>
          <div class="flex font-bold items-center justify-between p-2 text-[#333333] text-xs/[16px] w-full">
            <p>
              2
            </p>
            <p>
              USD 12.50
            </p>
          </div>
        </div>
        <div style="border-top: 2px solid rgba(186, 243, 230, 1); margin-top: 20px;"></div>
      </div>
      <div class="py-[10px]">
        <div class="flex space-x-8">
          <div class="flex space-x-3">
            <img alt="Item 2" class="h-[45px] rounded-[22.5px] w-[45px]" src="https://via.placeholder.com/45">
            <div class="w-[225px]">
              <p class="font-normal text-[#004d5a] text-xs/[16px] underline">
                Item 2
              </p>
              <p class="mt-[4px] opacity-90 space-x-1 text-[#999999] text-[10px]/[16px]">
                USD 23.99
              </p>
            </div>
          </div>
          <div class="flex font-bold items-center justify-between p-2 text-[#333333] text-xs/[16px] w-full">
            <p>
              1
            </p>
            <p>
              USD 23.99
            </p>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</ey-items-card>
`;

exports[`ey-items-card should render correctly with arrayData as an object 1`] = `
<ey-items-card>
  <mock:shadow-root>
    <div class="bg-[#baf3e6] bg-opacity-10 border-[#BAF3E626] border-[2px] px-[8px] rounded-[6px] w-full">
      <div class="py-[10px]">
        <div class="flex space-x-8">
          <div class="flex space-x-3">
            <img alt="Item 1" class="h-[45px] rounded-[22.5px] w-[45px]" src="https://via.placeholder.com/45">
            <div class="w-[225px]">
              <p class="font-normal text-[#004d5a] text-xs/[16px] underline">
                Item 1
              </p>
              <p class="mt-[4px] opacity-90 space-x-1 text-[#999999] text-[10px]/[16px]">
                USD 12.5
              </p>
            </div>
          </div>
          <div class="flex font-bold items-center justify-between p-2 text-[#333333] text-xs/[16px] w-full">
            <p>
              2
            </p>
            <p>
              USD 12.50
            </p>
          </div>
        </div>
        <div style="border-top: 2px solid rgba(186, 243, 230, 1); margin-top: 20px;"></div>
      </div>
      <div class="py-[10px]">
        <div class="flex space-x-8">
          <div class="flex space-x-3">
            <img alt="Item 2" class="h-[45px] rounded-[22.5px] w-[45px]" src="https://via.placeholder.com/45">
            <div class="w-[225px]">
              <p class="font-normal text-[#004d5a] text-xs/[16px] underline">
                Item 2
              </p>
              <p class="mt-[4px] opacity-90 space-x-1 text-[#999999] text-[10px]/[16px]">
                USD 23.99
              </p>
            </div>
          </div>
          <div class="flex font-bold items-center justify-between p-2 text-[#333333] text-xs/[16px] w-full">
            <p>
              1
            </p>
            <p>
              USD 23.99
            </p>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</ey-items-card>
`;
